env/context.go

```diff
@@ -5,59 +5,174 @@
+func (c *KBFSContext) GetLogDir() string {
 	return c.g.Env.GetLogDir()
 }
 
 // GetDataDir returns log dir
-func (c Context) GetDataDir() string {
+func (c *KBFSContext) GetDataDir() string {
 	return c.g.Env.GetDataDir()
 }
 
 // GetRunMode returns run mode
-func (c Context) GetRunMode() libkb.RunMode {
+func (c *KBFSContext) GetRunMode() libkb.RunMode {
 	return c.g.GetRunMode()
 }
 
 // GetSocket returns a socket
-func (c Context) GetSocket(clearError bool) (net.Conn, rpc.Transporter, bool, error) {
+func (c *KBFSContext) GetSocket(clearError bool) (
+	net.Conn, rpc.Transporter, bool, error) {
 	return c.g.GetSocket(clearError)
 }
 
 // ConfigureSocketInfo configures a socket
-func (c Context) ConfigureSocketInfo() error {
+func (c *KBFSContext) ConfigureSocketInfo() error {
 	return c.g.ConfigureSocketInfo()
 }
 
 // NewRPCLogFactory constructs an RPC logger
-func (c Context) NewRPCLogFactory() *libkb.RPCLogFactory {
+func (c *KBFSContext) NewRPCLogFactory() *libkb.RPCLogFactory {
 	return &libkb.RPCLogFactory{Contextified: libkb.NewContextified(c.g)}
 }
+
+func (c *KBFSContext) getSandboxSocketFile() string {
+	sandboxDir := c.g.Env.HomeFinder.SandboxCacheDir()
+	if sandboxDir == "" {
+		return ""
+	}
+	return filepath.Join(sandboxDir, kbfsSocketFile)
+}
+
+func (c *KBFSContext) getkbfsSocketFile() string {
+	e := c.g.Env
+	return e.GetString(
+		func() string { return c.getSandboxSocketFile() },
+		// TODO: maybe add command-line option here
+		func() string { return os.Getenv("KBFS_SOCKET_FILE") },
+		func() string { return filepath.Join(e.GetRuntimeDir(), kbfsSocketFile) },
+	)
+}
+
+func (c *KBFSContext) newTransportFromSocket(s net.Conn) rpc.Transporter {
+	return rpc.NewTransport(s, c.NewRPCLogFactory(), libkb.WrapError)
+}
+
+// GetKBFSSocket dials the socket configured in `c.kbfsSocket`.
+func (c *KBFSContext) GetKBFSSocket(clearError bool) (
+	net.Conn, rpc.Transporter, bool, error) {
+
+	var err error
+	c.g.Trace("GetSocket", func() error { return err })()
+
+	// Protect all global socket wrapper manipulation with a
+	// lock to prevent race conditions.
+	c.kbfsSocketMtx.Lock()
+	defer c.kbfsSocketMtx.Unlock()
+
+	needWrapper := false
+	if c.kbfsSocketWrapper == nil {
+		needWrapper = true
+		c.g.Log.Debug("| empty socket wrapper; need a new one")
+	} else if c.kbfsSocketWrapper.Transporter != nil && !c.kbfsSocketWrapper.Transporter.IsConnected() {
+		// need reconnect
+		c.g.Log.Debug("| rpc transport isn't connected, reconnecting...")
+		needWrapper = true
+	}
+
+	isNew := false
+	if needWrapper {
+		sw := libkb.SocketWrapper{}
+		if c.kbfsSocket == nil {
+			sw.Err = fmt.Errorf("Cannot get socket")
+		} else {
+			sw.Conn, sw.Err = c.kbfsSocket.DialSocket()
+			c.g.Log.Debug("| DialSocket -> %s", libkb.ErrToOk(sw.Err))
+			isNew = true
+		}
+		if sw.Err == nil {
+			sw.Transporter = c.newTransportFromSocket(sw.Conn)
+		}
+		c.kbfsSocketWrapper = &sw
+	}
+
+	// Return the current error no matter what
```

Comment:
I notice the original code has a `err = sw.err` before after this block.  I guess the comment is referring to the fact that we don't do that. Can you add more details about why?

Reply:
Hmm I think we actually should have it.